### PR TITLE
Big WebSockets refactor

### PR DESCRIPTION
**Pull request by chelog**
There's a result of my thoughts I commented on #11:

- Made auth token refresh while expiring instead of constant full reconnects, which may miss some data
- Added `Shard.refreshToken()` method
- `Shard.connect()`, `Shard.reconnect()` and `Shard.refreshToken()` new return promises resolved after successful auth
- Removed `ws` option from config and made all socket connections manual with `client.addSocketServer().connect()`
- Moved events from PteroClient to Shard, added typings for them
- Added typings for `Shard.send()` commands and `ServerStats` from `statsUpdate` event
- Documented most of `Shard` methods
- Fixed minor stuff I stumbled upon (ping, readyAt and such)

Open to any discussions, really prefer if you'd pinged me on Discord instead of comments here :)

---

Test code:
```ts
import { PteroClient, ServerStats } from './PteroJS';

const wait = (delay: number) => new Promise(resolve => setTimeout(resolve, delay))

async function run() {
	const client = new PteroClient('-snip-', '-snip-')
	await client.connect()

	const shard = client.addSocketServer('d7d01832')
	shard.on('serverConnect', socket => console.log(`socket connected for ${shard.id}: ${socket.url}`))
	shard.on('serverDisconnect', () => console.log(`socket disconnected for ${shard.id}`))

	const socket = await shard.connect()

	console.log('listening stats for 3 seconds...')
	const logStats = (stats: ServerStats) => console.log(`${shard.id}'s cpu: ${stats.cpu_absolute}`)
	shard.on('statsUpdate', logStats)
	await wait(3000)
	shard.off('statsUpdate', logStats)

	// dont't forget to reset console colors
	shard.on('serverOutput', (output) => console.log(`${shard.id}'s output: ${output}\x1b[0m`))
	shard.send('send command', ['hello'])

	await shard.disconnect()
	await shard.connect()

	shard.send('send command', ['hello after reconnection'])
}

run()
```

Result:

### CODE REVIEW

File: src/client/ws/packetHandler.js [hunk 1]
```diff
@@ -1,48 +1,44 @@
-function handle(client, { event, args }, id) {
-    function getServer(id) {
```
Comment (devnote-dev) [inline src/client/ws/packetHandler.js:2]: Why was the `getServer()` function removed here?

File: src/index.d.ts [hunk 2]
```diff
@@ -235,26 +235,80 @@ export type WebSocketStatus =
...
-export class Shard {
+export interface ShardCommands {
+    'auth': [token: string]
+    'send stats': undefined
+    'send logs': undefined
+    'set state': [state: PowerState]
+    'send command': [command: string]
+}
+
+export interface ServerStats {
```
Comment (devnote-dev) [inline src/index.d.ts:246]: The object keys here should be in camelCase (can be changed with caseConv util).

File: src/index.d.ts [hunk 3]
```diff
@@ -235,26 +235,80 @@ export type WebSocketStatus =
...
+    memory_limit_bytes: number;
+    network: {
+        rx_bytes: number;
+        tx_bytes: number;
+    };
+    state: string;
+    uptime: number;
+}
+
+export interface ShardEvents {
```
Comment (devnote-dev) [inline src/index.d.ts:259]: Some of the events here should include the server too.

**Review by devnote-dev — CHANGES_REQUESTED** [review]
Overall these are good changes but I have a few concerns

File: src/client/ws/packetHandler.js (hunk 1 above)
Comment (chelog) [reply src/client/ws/packetHandler.js:2]: You don't have servers in cache if you only use sockets. Accessing cache at every socket event is inefficient, imagine every server console message on busy game or web servers

File: src/index.d.ts (hunk 2 above)
Comment (chelog) [reply src/index.d.ts:246]: These strings are from panel api. Do you suggest creating a map and "translate" commands for panel?

Comment (devnote-dev) [reply src/index.d.ts:246]: `src/util/caseConv.js` can be used for conversion the packetHandler file.

Comment (chelog) [reply src/index.d.ts:246]: Sure, will do in a moment

**Review by devnote-dev — APPROVED** [review]

**Pull request merged**